gridcomps/History3G/HistoryCollectionGridComp_private.F90

```diff
@@ -33,31 +36,71 @@ function make_geom(hconfig, rc) result(geom)
       _RETURN(_SUCCESS)
    end function make_geom
 
-   !subroutine make_import_state(gridcomp, hconfig, rc)
-      !type(ESMF_GridComp), intent(inout) :: gridcomp
-      !type(ESMF_HConfig), intent(in) :: hconfig
-      !integer, optional, intent(out) :: rc
+   subroutine register_imports(gridcomp, hconfig, rc)
+      type(ESMF_GridComp), intent(inout) :: gridcomp
+      type(ESMF_HConfig), intent(in) :: hconfig
+      integer, optional, intent(out) :: rc
+      character(len=*), parameter :: VAR_LIST_KEY = 'var_list'
+      type(ESMF_HConfigIter) :: iter, iter_begin, iter_end
+      type(ESMF_HConfig) :: var_list
+      character(len=:), allocatable :: item_name
+      character(len=:), allocatable :: short_name
+      type(VariableSpec) :: varspec
+      integer :: status
+
+      var_list = ESMF_HConfigCreateAt(hconfig, keystring=VAR_LIST_KEY, _RC)
+      iter_begin = ESMF_HConfigIterBegin(var_list,_RC)
+      iter_end = ESMF_HConfigIterEnd(var_list,_RC)
+      iter = iter_begin
+
+      do while (ESMF_HConfigIterLoop(iter,iter_begin,iter_end,rc=status))
+         call parse_item(iter, item_name, short_name, _RC)
+         varspec = VariableSpec(ESMF_STATEINTENT_IMPORT, short_name)
+         call MAPL_AddSpec(gridcomp, varspec, _RC)
+      end do
+
+      _RETURN(_SUCCESS)
+   end subroutine register_imports
+
+   subroutine parse_item(item, item_name, short_name, rc)
+      type(ESMF_HConfigIter), intent(in) :: item 
+      character(len=:), allocatable, intent(out) :: item_name
+      character(len=:), allocatable, intent(out) :: short_name
+      integer, optional, intent(out) :: rc
+      character(len=*), parameter :: EXPRESSION_KEY = 'expr'
+      integer :: status
+      logical :: asOK, isScalar, isMap
+      type(ESMF_HConfig) :: value
+      type(ESMF_HConfigIter) :: iter, iterBegin, iterEnd
+      character(len=:), allocatable :: part_key, part_value
 
-      !type(ESMF_HConfigIter) :: iter, iter_begin, iter_end
-      !type(ESMF_HConfig) :: var_list
-      !character(len=:), allocatable :: var_name
-      !type(VariableSpec) :: varspec
-      !integer :: status
+      isScalar = ESMF_HConfigIsScalarMapKey(item, _RC)
+      _ASSERT(isScalar, 'Variable list item does not have a scalar name.')
 
-      !var_list = ESMF_HConfigCreateAt(hconfig, keystring='var_list', _RC)
-      !iter_begin = ESMF_HConfigIterBegin(var_list,_RC)
-      !iter_end = ESMF_HConfigIterEnd(var_list,_RC)
-      !iter = iter_begin
+      isMap = ESMF_HConfigIsMapMapVal(item, _RC)
```

Comment:
I'm a bit confused.  How can you assert that it is both a scalar and a map?

Reply:
The map key is Scalar (String). The map value is a Map. `item` is a Map (HConfig) [Map 1].  It maps the History Collection item name (alias) to a Map [Map 2]. Map 2 maps 'expr' to a String expression, which in the simplest case is a single import. For the simplest case, treating the Map 1 value as a String is simple, and it could be parsed to get the String following 'expr:'. In the future, an item in the collection could have attributes in addition to 'expr'. It was simpler to treat the Map 1 value as a Map, which avoids parsing and allows the additional attributes in the future.

Reply:
Ah.  Sorry.